File: internal/config/agentsrc.go
package config

import (
	"crypto/sha256"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"time"
)

// isDirEntry reports whether the path is a directory, following symlinks.
func isDirEntry(path string) bool {
	info, err := os.Stat(path)
	return err == nil && info.IsDir()
}

// StringsOrBool holds either a boolean flag (all/none) or a named list.
// It marshals/unmarshals as either a JSON bool or a JSON string array:
//
//	true             → All resources of this type
//	false            → No resources
//	["name1","name2"] → Only the named resources
type StringsOrBool struct {
	All   bool
	Names []string
}

// IsEnabled returns true if any resources are enabled (All or at least one name).
func (s StringsOrBool) IsEnabled() bool {
	return s.All || len(s.Names) > 0
}

// Contains returns true if name is covered (either All=true or name is in Names).
func (s StringsOrBool) Contains(name string) bool {
	if s.All {
		return true
	}
	for _, n := range s.Names {
		if n == name {
			return true
		}
	}
	return false
}

// Add appends name to Names unless All is true (already covers everything).
func (s *StringsOrBool) Add(name string) {
	if s.All {
		return
	}
	for _, n := range s.Names {
		if n == name {
			return // already present
		}
	}
	s.Names = append(s.Names, name)
}

// Remove removes name from Names. No-op if All is true.
func (s *StringsOrBool) Remove(name string) {
	if s.All {
		return
	}
	out := s.Names[:0]
	for _, n := range s.Names {
		if n != name {
			out = append(out, n)
		}
	}
	s.Names = out
}

func (s StringsOrBool) MarshalJSON() ([]byte, error) {
	if len(s.Names) > 0 {
		return json.Marshal(s.Names)
	}
	return json.Marshal(s.All)
}

func (s *StringsOrBool) UnmarshalJSON(data []byte) error {
	var b bool
	if err := json.Unmarshal(data, &b); err == nil {
		s.All = b
		s.Names = nil
		return nil
	}
	var names []string
	if err := json.Unmarshal(data, &names); err != nil {
		return fmt.Errorf("hooks/mcp field must be bool or string array: %w", err)
	}
	s.All = false
	s.Names = names
	return nil
}

// AgentsRCKGBridge is the bridge sub-config within the KG section.
type AgentsRCKGBridge struct {
	Enabled        bool     `json:"enabled"`
	AllowedIntents []string `json:"allowed_intents,omitempty"`
}

// AgentsRCKG is the knowledge-graph configuration block in agentsrc.json.
type AgentsRCKG struct {
	// GraphHome overrides KG_HOME env var for this project. Defaults to ~/.knowledge-graph.
	GraphHome string `json:"graph_home,omitempty"`
	// Backend selects the storage backend: "sqlite" (default) or "postgres".
	// Postgres requires KG_POSTGRES_URL.
	Backend string `json:"backend,omitempty"`
	// Bridge configures workflow/kg bridge query behaviour for this project.
	Bridge AgentsRCKGBridge `json:"bridge"`
}

// AgentsRC represents the .agentsrc.json manifest committed to a project repo.
type AgentsRC struct {
	Schema   string           `json:"$schema,omitempty"`
	Version  int              `json:"version"`
	Project  string           `json:"project,omitempty"`
	Skills   []string         `json:"skills,omitempty"`
	Rules    []string         `json:"rules,omitempty"`
	Agents   []string         `json:"agents,omitempty"`
	Hooks    StringsOrBool    `json:"hooks"`
	MCP      StringsOrBool    `json:"mcp"`
	Settings bool             `json:"settings"`
	Sources  []Source         `json:"sources"`
	KG       *AgentsRCKG      `json:"kg,omitempty"`
	Refresh  *RefreshMetadata `json:"refresh,omitempty"`

	// ExtraFields captures unknown JSON keys so Save() can round-trip them
	// instead of silently dropping legacy or custom fields.
	ExtraFields map[string]json.RawMessage `json:"-"`
}

// RefreshMetadata records the latest dot-agents install/refresh that updated a project.
type RefreshMetadata struct {
	Version     string `json:"version,omitempty"`
	Commit      string `json:"commit,omitempty"`
	Describe    string `json:"describe,omitempty"`
	RefreshedAt string `json:"refreshedAt,omitempty"`
}

// SetRefreshMetadata stores the latest refresh details in the manifest.
func (a *AgentsRC) SetRefreshMetadata(version, commit, describe string, refreshedAt time.Time) {
	if a == nil {
		return
	}
	a.Refresh = &RefreshMetadata{
		Version:     version,
		Commit:      commit,
		Describe:    describe,
		RefreshedAt: refreshedAt.UTC().Format(time.RFC3339),
	}
}

// agentsRCKnown lists all JSON keys owned by AgentsRC's known fields.
var agentsRCKnown = map[string]bool{
	"$schema": true, "version": true, "project": true,
	"skills": true, "rules": true, "agents": true,
	"hooks": true, "mcp": true, "settings": true, "sources": true,
	"kg": true, "refresh": true,
}

// agentsRCCore is an alias used in custom marshal/unmarshal to avoid
// infinite recursion while still using the standard json encoder.
type agentsRCCore struct {
	Schema   string           `json:"$schema,omitempty"`
	Version  int              `json:"version"`
	Project  string           `json:"project,omitempty"`
	Skills   []string         `json:"skills,omitempty"`
	Rules    []string         `json:"rules,omitempty"`
	Agents   []string         `json:"agents,omitempty"`
	Hooks    StringsOrBool    `json:"hooks"`
	MCP      StringsOrBool    `json:"mcp"`
	Settings bool             `json:"settings"`
	Sources  []Source         `json:"sources"`
	KG       *AgentsRCKG      `json:"kg,omitempty"`
	Refresh  *RefreshMetadata `json:"refresh,omitempty"`
}

func (a *AgentsRC) UnmarshalJSON(data []byte) error {
	var core agentsRCCore
	if err := json.Unmarshal(data, &core); err != nil {
		return err
	}
	a.Schema = core.Schema
	a.Version = core.Version
	a.Project = core.Project
	a.Skills = core.Skills
	a.Rules = core.Rules
	a.Agents = core.Agents
	a.Hooks = core.Hooks
	a.MCP = core.MCP
	a.Settings = core.Settings
	a.Sources = core.Sources
	a.KG = core.KG
	a.Refresh = core.Refresh

	var all map[string]json.RawMessage
	if err := json.Unmarshal(data, &all); err != nil {
		return err
	}
	for k, v := range all {
		if !agentsRCKnown[k] {
			if a.ExtraFields == nil {
				a.ExtraFields = make(map[string]json.RawMessage)
			}
			a.ExtraFields[k] = v
		}
	}
	return nil
}

func (a AgentsRC) MarshalJSON() ([]byte, error) {
	core := agentsRCCore{
		Schema:   a.Schema,
		Version:  a.Version,
		Project:  a.Project,
		Skills:   a.Skills,
		Rules:    a.Rules,
		Agents:   a.Agents,
		Hooks:    a.Hooks,
		MCP:      a.MCP,
		Settings: a.Settings,
		Sources:  a.Sources,
		KG:       a.KG,
		Refresh:  a.Refresh,
	}
	data, err := json.Marshal(core)
	if err != nil {
		return nil, err
	}
	if len(a.ExtraFields) == 0 {
		return data, nil
	}
	var m map[string]json.RawMessage
	if err := json.Unmarshal(data, &m); err != nil {
		return nil, err
	}
	for k, v := range a.ExtraFields {
		if _, exists := m[k]; !exists {
			m[k] = v
		}
	}
	return json.Marshal(m)
}

// Source describes where to find agent resources.
type Source struct {
	Type string `json:"type"`           // "local" | "git"
	Path string `json:"path,omitempty"` // override path for "local"
	URL  string `json:"url,omitempty"`  // repository URL for "git"
	Ref  string `json:"ref,omitempty"`  // branch/tag for "git"
}

const AgentsRCFile = ".agentsrc.json"

// LoadAgentsRC reads .agentsrc.json from the given project directory.
func LoadAgentsRC(projectPath string) (*AgentsRC, error) {
	path := filepath.Join(projectPath, AgentsRCFile)
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	var rc AgentsRC
	if err := json.Unmarshal(data, &rc); err != nil {
		return nil, fmt.Errorf("parsing %s: %w", AgentsRCFile, err)
	}
	// Default to a local source if none declared
	if len(rc.Sources) == 0 {
		rc.Sources = []Source{{Type: "local"}}
	}
	return &rc, nil
}

// Save writes the manifest to .agentsrc.json in projectPath.
func (a *AgentsRC) Save(projectPath string) error {
	path := filepath.Join(projectPath, AgentsRCFile)
	data, err := json.MarshalIndent(a, "", "  ")
	if err != nil {
		return fmt.Errorf("marshaling %s: %w", AgentsRCFile, err)
	}
	data = append(data, '\n')
	return os.WriteFile(path, data, 0644)
}

// AgentsCacheDir returns the root directory for cached remote sources.
func AgentsCacheDir() string {
	cacheHome := os.Getenv("XDG_CACHE_HOME")
	if cacheHome == "" {
		home, _ := os.UserHomeDir()
		cacheHome = filepath.Join(home, ".cache")
	}
	return filepath.Join(cacheHome, "dot-agents")
}

// GitSourceCacheDir returns the cache directory for a given git URL.
func GitSourceCacheDir(url string) string {
	hash := fmt.Sprintf("%x", sha256.Sum256([]byte(url)))[:12]
	return filepath.Join(AgentsCacheDir(), "sources", hash)
}

// AppendUnique appends s to slice only if not already present.
func AppendUnique(slice []string, s string) []string {
	for _, v := range slice {
		if v == s {
			return slice
		}
	}
	return append(slice, s)
}

// GenerateAgentsRC inspects ~/.agents/ and builds a manifest for the given project.
func GenerateAgentsRC(projectName, projectPath string) (*AgentsRC, error) {
	agentsHome := AgentsHome()

	rc := &AgentsRC{
		Schema:  "https://dot-agents.dev/schemas/agentsrc.json",
		Version: 1,
		Project: projectName,
		Sources: []Source{{Type: "local"}},
	}

	scopes := []string{"global", projectName}
	rc.Skills = collectScopedDirs(agentsHome, "skills", scopes, "SKILL.md")
	rc.Agents = collectScopedDirs(agentsHome, "agents", scopes, "AGENT.md")
	rc.Rules = detectRuleScopes(agentsHome, projectName)
	rc.Hooks = detectHookEvents(agentsHome, projectName)
	rc.MCP = detectMCPServers(agentsHome, projectName)
	rc.Settings = detectPlatformSettings(agentsHome, projectName)

	return rc, nil
}

// MergeGenerateAgentsRC overlays a freshly generated manifest onto an existing
// on-disk manifest. Scan-derived lists (skills, rules, agents, hooks, mcp,
// settings) come from generated; an existing non-empty project name, unknown
// JSON keys (ExtraFields), and supplemental sources (e.g. git remotes not
// produced by GenerateAgentsRC) are preserved. Source entries are unioned with
// deduplication so the default local source is not duplicated when merging.
func MergeGenerateAgentsRC(existing, generated *AgentsRC) *AgentsRC {
	if existing == nil {
		return generated
	}
	if generated == nil {
		return existing
	}
	out := *generated
	out.Sources = mergeSourceSlices(generated.Sources, existing.Sources)
	if existing.Project != "" {
		out.Project = existing.Project
	}
	if len(existing.ExtraFields) > 0 {
		out.ExtraFields = cloneExtraFieldsMap(existing.ExtraFields)
	}
	if existing.Refresh != nil {
		out.Refresh = existing.Refresh
	}
	return &out
}

func cloneExtraFieldsMap(m map[string]json.RawMessage) map[string]json.RawMessage {
	out := make(map[string]json.RawMessage, len(m))
	for k, v := range m {
		out[k] = v
	}
	return out
}

func mergeSourceSlices(generated, existing []Source) []Source {
	seen := make(map[string]bool)
	var out []Source
	for _, s := range generated {
		k := sourceMergeKey(s)
		if seen[k] {
			continue
		}
		seen[k] = true
		out = append(out, s)
	}
	for _, s := range existing {
		k := sourceMergeKey(s)
		if seen[k] {
			continue
		}
		seen[k] = true
		out = append(out, s)
	}
	return out
}

func sourceMergeKey(s Source) string {
	switch s.Type {
	case "local":
		return "local:" + s.Path
	case "git":
		return "git:" + s.URL + "\x00" + s.Ref
	default:
		return "type:" + s.Type + "\x00" + s.Path + "\x00" + s.URL + "\x00" + s.Ref
	}
}

// collectScopedDirs returns unique entry names from resource subdirs that contain markerFile.
func collectScopedDirs(agentsHome, resourceType string, scopes []string, markerFile string) []string {
	var names []string
	for _, scope := range scopes {
		dir := filepath.Join(agentsHome, resourceType, scope)
		entries, err := os.ReadDir(dir)
		if err != nil {
			continue
		}
		for _, e := range entries {
			entryPath := filepath.Join(dir, e.Name())
			if !isDirEntry(entryPath) {
				continue
			}
			if _, err := os.Stat(filepath.Join(entryPath, markerFile)); err == nil {
				names = AppendUnique(names, e.Name())
			}
		}
	}
	return names
}

// detectHookEvents reads the project claude-code.json and returns a StringsOrBool
// listing hook event names that have at least one entry.
func detectHookEvents(agentsHome, projectName string) StringsOrBool {
	for _, scope := range []string{projectName, "global"} {
		hooksDir := filepath.Join(agentsHome, "hooks", scope)
		entries, err := os.ReadDir(hooksDir)
		if err == nil {
			for _, entry := range entries {
				if !entry.IsDir() {
					continue
				}
				if _, err := os.Stat(filepath.Join(hooksDir, entry.Name(), "HOOK.yaml")); err == nil {
					return StringsOrBool{All: true}
				}
			}
		}
	}

	for _, scope := range []string{projectName, "global"} {
		settingsPath := filepath.Join(agentsHome, "settings", scope, "claude-code.json")
		data, err := os.ReadFile(settingsPath)
		if err != nil {
			continue
		}
		var settings map[string]any
		if json.Unmarshal(data, &settings) != nil {
			continue
		}
		hooksVal, ok := settings["hooks"]
		if !ok {
			continue
		}
		hooksMap, ok := hooksVal.(map[string]any)
		if !ok {
			continue
		}
		var hookEvents []string
		for event, val := range hooksMap {
			if list, ok := val.([]any); ok && len(list) > 0 {
				hookEvents = append(hookEvents, event)
			}
		}
		if len(hookEvents) == 0 {
			continue
		}
		sort.Strings(hookEvents)
		return StringsOrBool{Names: hookEvents}
	}
	return StringsOrBool{}
}

// detectMCPServers scans MCP config files for the project and global scopes
// and returns a StringsOrBool listing named server entries.
func detectMCPServers(agentsHome, projectName string) StringsOrBool {
	for _, scope := range []string{projectName, "global"} {
		if result := readMCPScope(agentsHome, scope); result.IsEnabled() {
			return result
		}
	}
	return StringsOrBool{}
}

// readMCPScope tries claude.json, mcp.json, then .mcp.json for a single scope directory
// and returns the server list from the first readable file.
func readMCPScope(agentsHome, scope string) StringsOrBool {
	for _, fname := range []string{"claude.json", "mcp.json", ".mcp.json"} {
		mcpPath := filepath.Join(agentsHome, "mcp", scope, fname)
		data, err := os.ReadFile(mcpPath)
		if err != nil {
			continue
		}
		var mcpConfig map[string]any
		if json.Unmarshal(data, &mcpConfig) != nil {
			continue
		}
		servers, ok := mcpConfig["servers"].(map[string]any)
		if !ok {
			servers, ok = mcpConfig["mcpServers"].(map[string]any)
		}
		if !ok {
			break
		}
		var names []string
		for name := range servers {
			names = append(names, name)
		}
		if len(names) > 0 {
			sort.Strings(names)
			return StringsOrBool{Names: names}
		}
		break // found a file, stop trying filenames
	}
	return StringsOrBool{}
}

// detectPlatformSettings returns true if a cursor.json settings file exists
// for the project or global scope.
func detectPlatformSettings(agentsHome, projectName string) bool {
	for _, scope := range []string{projectName, "global"} {
		if _, err := os.Stat(filepath.Join(agentsHome, "settings", scope, "cursor.json")); err == nil {
			return true
		}
	}
	return false
}

func detectRuleScopes(agentsHome, projectName string) []string {
	scopes := []string{"global"}
	projectRulesDir := filepath.Join(agentsHome, "rules", projectName)
	entries, err := os.ReadDir(projectRulesDir)
	if err != nil {
		return scopes
	}
	for _, entry := range entries {
		ext := filepath.Ext(entry.Name())
		if ext == ".md" || ext == ".mdc" || ext == ".txt" {
			return append(scopes, "project")
		}
	}
	return scopes
}

File: ports/typescript/src/core/agentsrc.ts
/**
 * Minimal .agentsrc.json handling aligned with internal/config/agentsrc.go:
 * known keys map into typed fields; other top-level keys are kept in extraFields
 * and merged back on serialize so they are not dropped (see TestAgentsRCUnknownFieldsRoundtrip).
 */

import { readFile, writeFile } from "node:fs/promises";
import { join } from "node:path";

export const AGENTS_RC_FILE = ".agentsrc.json";

/** Keys owned by AgentsRC in Go (agentsRCKnown). */
export const AGENTS_RC_KNOWN_KEYS = new Set([
  "$schema",
  "version",
  "project",
  "skills",
  "rules",
  "agents",
  "hooks",
  "mcp",
  "settings",
  "sources",
]);

export interface Source {
  type: string;
  path?: string;
  url?: string;
  ref?: string;
}

/** Mirrors config.StringsOrBool semantics (bool or string array in JSON). */
export interface StringsOrBool {
  all: boolean;
  names: string[];
}

export interface AgentsRc {
  schema?: string;
  version: number;
  project?: string;
  skills?: string[];
  rules?: string[];
  agents?: string[];
  hooks: StringsOrBool;
  mcp: StringsOrBool;
  settings: boolean;
  sources: Source[];
  /** Top-level JSON keys not in AGENTS_RC_KNOWN_KEYS — preserved on save. */
  extraFields: Record<string, unknown>;
}

function parseStringsOrBool(raw: unknown, field: string): StringsOrBool {
  if (typeof raw === "boolean") {
    return raw ? { all: true, names: [] } : { all: false, names: [] };
  }
  if (Array.isArray(raw) && raw.every((x) => typeof x === "string")) {
    return { all: false, names: raw as string[] };
  }
  throw new Error(`${field} must be a boolean or string array`);
}

function stringsOrBoolToJson(s: StringsOrBool): boolean | string[] {
  if (s.names.length > 0) {
    return s.names;
  }
  return s.all;
}

function readSources(raw: unknown): Source[] {
  if (raw === undefined) {
    return [];
  }
  if (!Array.isArray(raw)) {
    throw new Error("sources must be an array");
  }
  const out: Source[] = [];
  for (const item of raw) {
    if (item === null || typeof item !== "object" || Array.isArray(item)) {
      throw new Error("each source must be an object");
    }
    const o = item as Record<string, unknown>;
    const type = o.type;
    if (typeof type !== "string") {
      throw new Error('source must have string "type"');
    }
    const s: Source = { type };
    if (typeof o.path === "string") {
      s.path = o.path;
    }
    if (typeof o.url === "string") {
      s.url = o.url;
    }
    if (typeof o.ref === "string") {
      s.ref = o.ref;
    }
    out.push(s);
  }
  return out;
}

function readStringArray(raw: unknown, field: string): string[] | undefined {
  if (raw === undefined) {
    return undefined;
  }
  if (!Array.isArray(raw) || !raw.every((x) => typeof x === "string")) {
    throw new Error(`${field} must be an array of strings`);
  }
  return raw as string[];
}

/** Parse JSON text into AgentsRc, splitting unknown top-level keys into extraFields. */
export function parseAgentsRcJson(text: string): AgentsRc {
  let root: unknown;
  try {
    root = JSON.parse(text) as unknown;
  } catch (e) {
    throw new Error(`parsing ${AGENTS_RC_FILE}: ${e}`);
  }
  if (root === null || typeof root !== "object" || Array.isArray(root)) {
    throw new Error(`${AGENTS_RC_FILE} must contain a JSON object`);
  }
  const all = root as Record<string, unknown>;

  const extraFields: Record<string, unknown> = {};
  for (const key of Object.keys(all)) {
    if (!AGENTS_RC_KNOWN_KEYS.has(key)) {
      extraFields[key] = all[key];
    }
  }

  const version = all.version;
  if (typeof version !== "number" || !Number.isInteger(version)) {
    throw new Error("version must be an integer");
  }

  const project = all.project;
  if (project !== undefined && typeof project !== "string") {
    throw new Error("project must be a string");
  }

  const schema = all.$schema;
  if (schema !== undefined && typeof schema !== "string") {
    throw new Error("$schema must be a string");
  }

  let sources = readSources(all.sources);
  if (sources.length === 0) {
    sources = [{ type: "local" }];
  }

  const hooks =
    all.hooks === undefined ? { all: false, names: [] } : parseStringsOrBool(all.hooks, "hooks");
  const mcp = all.mcp === undefined ? { all: false, names: [] } : parseStringsOrBool(all.mcp, "mcp");

  const rc: AgentsRc = {
    schema,
    version,
    project,
    skills: readStringArray(all.skills, "skills"),
    rules: readStringArray(all.rules, "rules"),
    agents: readStringArray(all.agents, "agents"),
    hooks,
    mcp,
    settings: Boolean(all.settings),
    sources,
    extraFields,
  };

  return rc;
}

type CorePlain = {
  $schema?: string;
  version: number;
  project?: string;
  skills?: string[];
  rules?: string[];
  agents?: string[];
  hooks: boolean | string[];
  mcp: boolean | string[];
  settings: boolean;
  sources: Source[];
};

function toCorePlain(rc: AgentsRc): CorePlain {
  const core: CorePlain = {
    version: rc.version,
    hooks: stringsOrBoolToJson(rc.hooks),
    mcp: stringsOrBoolToJson(rc.mcp),
    settings: rc.settings,
    sources: rc.sources,
  };
  if (rc.schema !== undefined) {
    core.$schema = rc.schema;
  }
  if (rc.project !== undefined) {
    core.project = rc.project;
  }
  if (rc.skills !== undefined && rc.skills.length > 0) {
    core.skills = rc.skills;
  }
  if (rc.rules !== undefined && rc.rules.length > 0) {
    core.rules = rc.rules;
  }
  if (rc.agents !== undefined && rc.agents.length > 0) {
    core.agents = rc.agents;
  }
  return core;
}

/**
 * Serialize AgentsRc to formatted JSON (2-space indent, trailing newline), matching Go Save/MarshalIndent.
 * Unknown keys are merged like AgentsRC.MarshalJSON: only add an extra key if the core object does not already define it.
 */
export function serializeAgentsRc(rc: AgentsRc): string {
  const core = toCorePlain(rc);
  const data = JSON.stringify(core);
  const m = JSON.parse(data) as Record<string, unknown>;
  for (const [k, v] of Object.entries(rc.extraFields)) {
    if (!(k in m)) {
      m[k] = v;
    }
  }
  return `${JSON.stringify(m, null, 2)}\n`;
}

export async function loadAgentsRc(projectPath: string): Promise<AgentsRc> {
  const path = join(projectPath, AGENTS_RC_FILE);
  const text = await readFile(path, "utf8");
  return parseAgentsRcJson(text);
}

export async function saveAgentsRc(projectPath: string, rc: AgentsRc): Promise<void> {
  const path = join(projectPath, AGENTS_RC_FILE);
  await writeFile(path, serializeAgentsRc(rc), "utf8");
}
